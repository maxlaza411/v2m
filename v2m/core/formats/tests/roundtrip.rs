use std::fs::File;
use std::path::PathBuf;

use insta::assert_json_snapshot;
use serde_json::Value;
use v2m_formats as formats;

fn examples_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../examples")
}

#[test]
fn constraints_roundtrip_snapshot() {
    let path = examples_dir().join("constraints/minimal.json");
    let doc = formats::constraints::from_reader(File::open(&path).expect("open constraints"))
        .expect("load constraints");
    let mut buf = Vec::new();
    formats::constraints::to_writer(&doc, &mut buf).expect("write constraints");
    let doc2 = formats::constraints::from_reader(buf.as_slice()).expect("reload constraints");
    assert_eq!(doc, doc2);
    let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
    assert_json_snapshot!("constraints_roundtrip", value);
}

#[test]
fn nir_roundtrip_snapshot() {
    let fixtures = ["minimal.json", "full_adder.json", "counter.json"];
    let base_dir = examples_dir().join("nir");

    for fixture in fixtures {
        let path = base_dir.join(fixture);
        let doc =
            formats::nir::from_reader(File::open(&path).expect("open nir")).expect("load nir");
        let mut buf = Vec::new();
        formats::nir::to_writer(&doc, &mut buf).expect("write nir");
        let doc2 = formats::nir::from_reader(buf.as_slice()).expect("reload nir");
        assert_eq!(doc, doc2);
        let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
        let snapshot_name = format!(
            "nir_roundtrip__{}",
            fixture.strip_suffix(".json").unwrap_or(fixture)
        );
        assert_json_snapshot!(snapshot_name, value);
    }
}

#[test]
fn tir_roundtrip_snapshot() {
    let path = examples_dir().join("tir/minimal.json");
    let doc = formats::tir::from_reader(File::open(&path).expect("open tir")).expect("load tir");
    let mut buf = Vec::new();
    formats::tir::to_writer(&doc, &mut buf).expect("write tir");
    let doc2 = formats::tir::from_reader(buf.as_slice()).expect("reload tir");
    assert_eq!(doc, doc2);
    let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
    assert_json_snapshot!("tir_roundtrip", value);
}

#[test]
fn pir_roundtrip_snapshot() {
    let path = examples_dir().join("pir/minimal.json");
    let doc = formats::pir::from_reader(File::open(&path).expect("open pir")).expect("load pir");
    let mut buf = Vec::new();
    formats::pir::to_writer(&doc, &mut buf).expect("write pir");
    let doc2 = formats::pir::from_reader(buf.as_slice()).expect("reload pir");
    assert_eq!(doc, doc2);
    let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
    assert_json_snapshot!("pir_roundtrip", value);
}

#[test]
fn techlib_roundtrip_snapshot() {
    let path = examples_dir().join("techlib/minimal.json");
    let doc = formats::techlib::from_reader(File::open(&path).expect("open techlib"))
        .expect("load techlib");
    let mut buf = Vec::new();
    formats::techlib::to_writer(&doc, &mut buf).expect("write techlib");
    let doc2 = formats::techlib::from_reader(buf.as_slice()).expect("reload techlib");
    assert_eq!(doc, doc2);
    let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
    assert_json_snapshot!("techlib_roundtrip", value);
}

#[test]
fn wir_roundtrip_snapshot() {
    let path = examples_dir().join("wir/minimal.json");
    let doc = formats::wir::from_reader(File::open(&path).expect("open wir")).expect("load wir");
    let mut buf = Vec::new();
    formats::wir::to_writer(&doc, &mut buf).expect("write wir");
    let doc2 = formats::wir::from_reader(buf.as_slice()).expect("reload wir");
    assert_eq!(doc, doc2);
    let value: Value = serde_json::from_slice(&buf).expect("roundtrip value");
    assert_json_snapshot!("wir_roundtrip", value);
}
