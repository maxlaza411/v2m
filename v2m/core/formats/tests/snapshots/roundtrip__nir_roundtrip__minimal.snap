---
source: core/formats/tests/roundtrip.rs
expression: value
---
{
  "design": "example",
  "modules": {
    "Example": {
      "nets": {
        "a": {
          "bits": 1
        },
        "b": {
          "bits": 1
        },
        "n0": {
          "bits": 1
        },
        "y": {
          "bits": 1
        }
      },
      "nodes": {
        "and0": {
          "op": "AND",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 0,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 0,
              "net": "b"
            },
            "Y": {
              "lsb": 0,
              "msb": 0,
              "net": "y"
            }
          },
          "uid": "and0",
          "width": 1
        },
        "const0": {
          "op": "CONST",
          "pin_map": {
            "Y": {
              "const": "1",
              "width": 1
            }
          },
          "uid": "const0",
          "width": 1
        },
        "mux0": {
          "op": "MUX",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 0,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 0,
              "net": "b"
            },
            "S": {
              "concat": [
                {
                  "lsb": 0,
                  "msb": 0,
                  "net": "n0"
                }
              ]
            },
            "Y": {
              "lsb": 0,
              "msb": 0,
              "net": "y"
            }
          },
          "uid": "mux0",
          "width": 1
        }
      },
      "ports": {
        "a": {
          "bits": 1,
          "dir": "input"
        },
        "b": {
          "bits": 1,
          "dir": "input"
        },
        "y": {
          "bits": 1,
          "dir": "output"
        }
      }
    }
  },
  "top": "Example",
  "v": "nir-1.1"
}
